routers/repo/migrate.go

```diff
@@ -188,3 +193,53 @@ func MigratePost(ctx *context.Context, form auth.MigrateRepoForm) {
 
 	handleMigrateError(ctx, ctxUser, err, "MigratePost", tpl, &form)
 }
+
+// CancelMigration cancels a running migration
+func CancelMigration(ctx *context.Context) {
+	if !ctx.Repo.IsOwner() {
+		ctx.Error(http.StatusNotFound)
+		return
+	}
+
+	if ctx.Repo.Repository.Status != models.RepositoryBeingMigrated {
+		ctx.Error(http.StatusConflict, "repo already migrated")
+		return
+	}
+
+	guid := process.GetManager().GUID()
+	task, err := models.GetMigratingTask(ctx.Repo.Repository.ID)
+	if err != nil {
+		ctx.InternalServerError(err)
+		return
+	}
+	if task.Status != structs.TaskStatusRunning {
+		task.Status = structs.TaskStatusStopped
+		task.EndTime = timeutil.TimeStampNow()
+		task.RepoID = 0
+
+		if err = task.UpdateCols("status", "repo_id", "end_time"); err != nil {
+			log.Error("Task UpdateCols failed: %v", err)
+		}
+
+		if err := repo_service.DeleteRepository(ctx.User, ctx.Repo.Repository); err != nil {
+			ctx.ServerError("DeleteRepository", err)
+			return
+		}
+
+		ctx.Flash.Success(ctx.Tr("repo.migrate.cancelled"))
+		ctx.Redirect(ctx.Repo.Owner.DashboardLink())
+		return
+	}
+
+	tGUID, tPID := task.GUIDandPID()
+	if guid == tGUID {
```

Comment:
So once there is a running task and the gitea instance restarted. The tasks should be set stopped when the instance restarted. But since guid maybe changed, how do I know which running tasks should be stopped?

Reply:
since task is stopped, this doesn't mather - if it will be started again after a gitea-restart, it will update those values